doppler/provider_utils.go

```diff
@@ -0,0 +1,25 @@
+package doppler
+
+import (
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func handleNotFoundError(err error, d *schema.ResourceData) diag.Diagnostics {
+	if strings.Contains(err.Error(), "Could not find requested") {
+		// the resource no longer exists, so reset its ID so Terraform will
+		// generate a plan that recreates it
+		d.SetId("")
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  err.Error(),
+				Detail:   "Resource was not found, so it was removed from state and is being recreated.",
+			},
+		}
+	}
```

Comment:
+1 on @nmanoogian's point that we should rely on the status code, rather than the string error message (which could easily change in the future).

Reply:
I could be wrong, but from what I can tell, the `err` object we get doesn't have the status code. It literally just has the error string that you access by calling `Error()`.

Reply:
I suppose we could update `APIError` to include the status code and then swap everything over to using that?

Reply:
Even though we're returning the `error` interface, we're returning a concrete `APIError` under the hood. If you add the status code to `APIError`, you can check if the returned `error` is an instance of `APIError`. Likely something like this:

```diff
diff --git a/doppler/api.go b/doppler/api.go
index c227477..a94ba91 100644
--- a/doppler/api.go
+++ b/doppler/api.go
@@ -39,6 +39,7 @@ type APIError struct {
 	Err        error
 	Message    string
 	RetryAfter *time.Duration
+	Response   *APIResponse
 }
 
 type ErrorResponse struct {
@@ -140,16 +141,17 @@ func (client APIClient) PerformRequest(req *http.Request, params []QueryParam) (
 	defer r.Body.Close()
 
 	body, err := ioutil.ReadAll(r.Body)
+	response := &APIResponse{HTTPResponse: r, Body: body}
+
 	if err != nil {
-		return &APIResponse{HTTPResponse: r, Body: nil}, &APIError{Err: err, Message: "Unable to load response data"}
+		return &APIResponse{HTTPResponse: r, Body: nil}, &APIError{Err: err, Message: "Unable to load response data", Response: response}
 	}
-	response := &APIResponse{HTTPResponse: r, Body: body}
 
 	if !isSuccess(r.StatusCode) {
 		if contentType := r.Header.Get("content-type"); strings.HasPrefix(contentType, "application/json") {
 			var errResponse ErrorResponse
 			if err := json.Unmarshal(body, &errResponse); err != nil {
-				return response, &APIError{Err: err, Message: "Unable to load response"}
+				return response, &APIError{Err: err, Message: "Unable to load response", Response: response}
 			}
 
 			var retryAfter *time.Duration
@@ -177,12 +179,13 @@ func (client APIClient) PerformRequest(req *http.Request, params []QueryParam) (
 				Err:        nil,
 				Message:    strings.Join(errResponse.Messages, "
"),
 				RetryAfter: retryAfter,
+				Response:   response,
 			}
 		}
-		return nil, &APIError{Err: fmt.Errorf("0 status code; 0 bytes", r.StatusCode, len(body)), Message: "Unable to load response"}
+		return nil, &APIError{Err: fmt.Errorf("0 status code; 0 bytes", r.StatusCode, len(body)), Message: "Unable to load response", Response: response}
 	}
 	if err != nil {
-		return nil, &APIError{Err: err, Message: "Unable to parse response data"}
+		return nil, &APIError{Err: err, Message: "Unable to parse response data", Response: response}
 	}
 	return response, nil
 }
diff --git a/doppler/provider_utils.go b/doppler/provider_utils.go
index 0c942ca..6658791 100644
--- a/doppler/provider_utils.go
+++ b/doppler/provider_utils.go
@@ -1,14 +1,12 @@
 package doppler
 
 import (
-	"strings"
-
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
 func handleNotFoundError(err error, d *schema.ResourceData) diag.Diagnostics {
-	if strings.Contains(err.Error(), "Could not find requested") {
+	if apiError, ok := err.(*APIError); ok && apiError.Response != nil && apiError.Response.HTTPResponse.StatusCode == 404 {
 		// the resource no longer exists, so reset its ID so Terraform will
 		// generate a plan that recreates it
 		d.SetId("")

```

If you want to go this direction, I'd recommend putting the `api.go` changes into a separate commit 👍